libs/checkpoint/langgraph/checkpoint/serde/encrypted.py

```diff
@@ -0,0 +1,90 @@
+import os
+from typing import Any
+
+from langgraph.checkpoint.serde.base import CipherProtocol, SerializerProtocol
+from langgraph.checkpoint.serde.jsonplus import JsonPlusSerializer
+
+
+class EncryptedSerializer(SerializerProtocol):
+    """Serializer that encrypts and decrypts data using an encryption protocol."""
+
+    def __init__(
+        self, cipher: CipherProtocol, serde: SerializerProtocol = JsonPlusSerializer()
+    ) -> None:
+        self.cipher = cipher
+        self.serde = serde
+
+    def dumps(self, obj: Any) -> bytes:
+        return self.serde.dumps(obj)
+
+    def loads(self, data: bytes) -> Any:
+        return self.serde.loads(data)
+
+    def dumps_typed(self, obj: Any) -> tuple[str, bytes]:
+        """Serialize an object to a tuple (type, bytes) and encrypt the bytes."""
+        # serialize data
+        typ, data = self.serde.dumps_typed(obj)
+        # encrypt data
+        encrypted_data = self.cipher.encrypt(data)
+        # add cipher name to type
+        return f"{typ}+{self.cipher.name}", encrypted_data
+
+    def loads_typed(self, data: tuple[str, bytes]) -> Any:
+        enc_typ, enc_data = data
+        # unencrypted data
+        if "+" not in enc_typ:
+            return self.serde.loads_typed(data)
+        # verify cipher name
+        typ, cipher_name = enc_typ.split("+", 1)
+        if cipher_name != self.cipher.name:
+            raise ValueError(
+                f"Cipher mismatch: expected {self.cipher.name}, got {cipher_name}"
+            )
+        # decrypt data
+        decrypted_data = self.cipher.decrypt(enc_data)
+        # deserialize data
+        return self.serde.loads_typed((typ, decrypted_data))
+
+    @classmethod
+    def from_pycryptodome_aes(
+        cls, serde: SerializerProtocol = JsonPlusSerializer(), **kwargs: Any
+    ) -> "EncryptedSerializer":
+        """Create an EncryptedSerializer using AES encryption."""
+        try:
+            from Crypto.Cipher import AES  # type: ignore
+        except ImportError:
+            raise ImportError(
+                "Pycryptodome is not installed. Please install it with `pip install pycryptodome`."
+            ) from None
+
+        # check if AES key is provided
+        if "key" in kwargs:
+            key = kwargs.pop("key")
+        else:
+            key = os.getenvb(b"LANGGRAPH_AES_KEY")
+            if key is None:
+                raise ValueError("LANGGRAPH_AES_KEY environment variable is not set.")
+            if len(key) not in (16, 24, 32):
+                raise ValueError("LANGGRAPH_AES_KEY must be 16, 24, or 32 bytes long.")
+
+        # set default mode to EAX if not provided
+        if kwargs.get("mode") is None:
+            kwargs["mode"] = AES.MODE_EAX
+
+        class PycryptodomeAesCipher(CipherProtocol):
```

Comment:
not sure if makes sense to split this out from inner class if want to reuse this

Reply:
Will make the conditional imports a little messier, so I think will leave like this